### PR TITLE
change solid harmonic ordering from gaussian to standard

### PR DESCRIPTION
**Pull request by loriab**
## Description
Kudos to @andysim for pointing out in January the exact 13 lines of code to change for the vast preponderance of the gaussian to standard ordering switch.

As far as timeline, there's no particular reason this needs to be in for v1.6. But if we _could_ avoid hosting forever one fewer set of L2 packages for a single release, that'd be kind of nice. But fixing the output formats (i.e., fchk, qcsk, molden, fcidump) is likely to take some careful work.

If anyone would like to tackle a part of this, go for it.

## Todos

- [ ] current test status (pretty impressive, imo)
  - after g2g commit
    - ctest: `99% tests passed, 3 tests failed out of 487`
    - pytest: `12 failed, 2971 passed, 270 skipped, 269 xfailed`
  - after first solidharmonics.cc commit
    - ctest: ~`90% tests passed, 50 tests failed out of 487`~
    - pytest: ~`43 failed, 2911 passed, 208 skipped, 269 xfailed`~
- [ ] only Linux in CI b/c only it has an appropriate L2 conda package built

- categories to examine
  - [x] tdscf
  - [x] cubeprop
  - [ ] fchk
  - [ ] qcsk atomicresult.wavefunction
  - [x] mbis
  - [ ] molden
  - [ ] fcidump
  - [ ] nbo

- fails pytest
  - [x]  tdscf
  - [x] dft_block_schemes
  - [ ] mcmurchie_davidson_consistency_angmom (fails because ref data are deposited in Gaussian ordering)
  - [ ] test_rhf_fchk[df-rhf-dft)]
  - [ ] test_rhf_fchk[conv-rhf-dct]
  - [ ] test_rhf_fchk[df-rhf-omp2]
  - [ ] test_uhf_fchk[df-uhf-mp2]
  - [ ] test_rhf_fchk[conv-rhf-ccsd] 
  - [ ] test_uhf_fchk[df-uhf-dft]
  - [ ] test_uhf_fchk[conv-uhf-ccsd]
  - [ ] test_rhf_fchk[df-rhf-mp2]
  - [ ] test_uhf_fchk[df-uhf]
  - [ ] test_rhf_fchk[df-rhf)]
  - [ ] test_rhf_fchk[conv-rhf-cc2]

- fails ctest
  - [x]	103 - cubeprop (Failed)
  - [x]	104 - cubeprop-frontier (Failed)
  - [x]	123 - density-screen-2 (Timeout)
  - [x]	161 - dft-grad-lr1 (Failed)
  - [x]	162 - dft-grad-lr2 (Failed)
  - [x]	163 - dft-grad-lr3 (Failed)
  - [x]	168 - dft-b2plyp (Failed)
  - [x]	170 - dft-ghost (Failed)
  - [x]	171 - dft-grad-meta (Failed)
  - [x]	173 - dft-freq-analytic (Failed)
  - [x]	174 - dft-grad1 (Failed)
  - [x]	175 - dft-grad2 (Failed)
  - [x]	177 - dft-b3lyp (Failed)
  - [x]	178 - dft1 (Failed)
  - [x]	179 - dft-vv10 (Failed)
  - [x]	180 - dft1-alt (Failed)
  - [x]	181 - dft2 (Failed)
  - [x]	182 - dft3 (Failed)
  - [x]	183 - dft-omega (Failed)
  - [x]	226 - frac-ip-fitting (Failed)
  - [x]	228 - frac-traverse (Failed)
  - [x]	234 - mbis-1 (Failed)
  - [x]	235 - mbis-2 (Failed)
  - [x]	236 - mbis-3 (Failed)
  - [x]	237 - mbis-4 (Failed)
  - [x]	238 - mbis-5 (Failed)
  - [x]	239 - mbis-6 (Failed)
  - [ ]	245 - mints3 (Failed) (fails because ref data are deposited in Gaussian ordering)
  - [x]	287 - olccd2 (Failed)
  - [x]	302 - omp3-3 (Failed)
  - [x]	307 - opt-lindep-change (Failed)
  - [x]	321 - opt13 (Failed)
  - [x]	348 - pywrap-checkrun-rhf (Failed)
  - [x]	350 - pywrap-checkrun-uhf (Failed)
  - [x]	366 - sapt-dft-api (Failed)
  - [x]	367 - sapt-dft-lrc (Failed)
  - [x]	390 - scf-level-shift-rks (Failed)
  - [x]	397 - soscf-dft (Failed)
  - [x]	401 - sapt-dft1 (Failed)
  - [x]	402 - sapt-dft2 (Failed)
  - [x]	405 - dft-custom (Failed)
  - [x]	425 - cubeprop-esp (Failed)
  - [ ]	434 - scf-coverage (Failed)
  - [x]	435 - dft-custom-dhdf (Failed)
  - [x]	437 - dft-custom-mgga (Failed)
  - [x]	438 - dft-custom-gga (Failed)
  - [x]	446 - tdscf-3 (Failed)
  - [x]	447 - tdscf-4 (Failed)
  - [x]	449 - tdscf-6 (Failed)
  - [x]         459 - linK-2 (Timeout)
  - [ ]         324 - opt-irc-2 (Failed)

## Checklist
- [ ] Tests added for any new features
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [ ] Ready for review
- [ ] Ready for merge

### CODE REVIEW

**Comment by andysim** [comment]
The DFT failures are almost certainly because of Gau2Grid using the wrong ordering convention.  Thankfully [CCA is supported](https://github.com/dgasmith/gau2grid/blob/master/docs/source/order.rst#spherical-order) so it should be fixable by a simple reconfiguration.

**Comment by maxscheurer** [comment]
For g2g, maybe just [this](https://github.com/psi4/psi4/blob/d9093c75c71c2b33fbe86f32b25d138675ac22eb/psi4/src/psi4/libfock/points.cc#L767) line needs to be changed?

**Comment by andysim** [comment]
> For g2g, maybe just [this](https://github.com/psi4/psi4/blob/d9093c75c71c2b33fbe86f32b25d138675ac22eb/psi4/src/psi4/libfock/points.cc#L767) line needs to be changed?

That looks exactly right to me - good call.  Switching `GG_SPHERICAL_GAUSSIAN` to `GG_SPHERICAL_CCA` should be all that's needed for the DFT tests.

**Comment by maxscheurer** [comment]
> For g2g, maybe just [this](https://github.com/psi4/psi4/blob/d9093c75c71c2b33fbe86f32b25d138675ac22eb/psi4/src/psi4/libfock/points.cc#L767) line needs to be changed?

Confirmed to work locally. This fixes *a lot* of the failed tests above 🎉

File: psi4/src/psi4/libmints/writer.cc [hunk 1]
```diff
@@ -411,12 +411,48 @@ void FCHKWriter::write(const std::string &filename) {
...
+        // Expected:  0   +1   -1   +2   -2   +3   -3   +4   -4
+        /* 0 */ {0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0},
+        /* 1 */ {0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0},
+        /* 2 */ {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+        /* 3 */ {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0},
+        /* 4 */ {0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+        /* 5 */ {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0},
+        /* 6 */ {0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+        /* 7 */ {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0},
+        /* 8 */ {1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
```
Comment (andysim) [inline psi4/src/psi4/libmints/writer.cc:455]:
```suggestion
        /* 8 */ {1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}
```

**Comment by robparrishqc** [comment]
I don't comment here all that often anymore, which is bittersweet (and I very much miss both my colleagues on PSI4 and the Vortex). We have gone through my old notes on CCA (sketch at best, we know), and many nested references therein. For Cartesian AOs, CCA matters - this lets you use a single normalization coefficient for everything, and tolerate non-normalized off-diagonal cartesian tensor components (i.e, xx is normalized in D, xy is not normalized in D, xxx is normalized in F, xxy or xyz is not normalized in F). For spherical AOs, as far as we can tell, there is *no* CCA convention except that everything be normalized (as literally all codes do). The one "CCA" paper we can find on this topic itself glancingly cites an older Schlegel / Frisch paper (I think before the big happening), which itself is outdated by about > two dozen articles on solid harmonics. Everyone who codes integrals deals with basis ordering - what you shudder about is weird normalization factors (particularly non-diagonal ones), not rigid permutations. 

The message is this: I am not convinced there is a standard. So maybe not optimal to follow to follow an implied one. Show me a reference of a real, used standard, and Lightspeed and Tachyon will be there tomorrow. 

[report (7).pdf](https://github.com/psi4/psi4/files/8746880/report.7.pdf)
[jcc.20815.pdf](https://github.com/psi4/psi4/files/8746882/jcc.20815.pdf)
[sh.pdf](https://github.com/psi4/psi4/files/8746881/sh.pdf)

**Comment by robparrishqc** [comment]
Last amplification - the JCC above does say, buried in some text: 

Screen Shot 2022-05-20 at 8 37 52 PM

But I've never heard of anyone using that order - it is anti-lexical. And while I have nothing but respect for the really good people who wrote that paper, nobody uses that standard (that one has 25 citations, PSI4 has 2000).

**Pull request closed without merging**